src/core/database/database.service.ts

```diff
@@ -218,48 +222,86 @@ export class DatabaseService {
    */
   async sgUpdateProperties<TObject extends Resource>({
     object,
-    props,
     changes,
     nodevar,
   }: {
     session: Session;
     object: TObject;
-    props: ReadonlyArray<keyof TObject & string>;
     changes: { [Key in keyof TObject]?: UnwrapSecured<TObject[Key]> };
     nodevar: string;
   }) {
     return await this.updateProperties({
       type: upperFirst(nodevar),
       object,
-      props,
       changes,
     });
   }
 
+  // expecting the changes of the "simple" properties.
+  async getActualChanges<TResource extends ResourceShape<any>>(
+    oldObject: TResource['prototype'],
+    changes: Partial<Record<keyof TResource['prototype'] & string, any>>
+  ): Promise<typeof changes> {
+    const props = Object.keys(changes);
+    // have a reference to 'input' most of the time, so need to copy the data instead of
+    //    accidentally mutating 'input'
+    const realChanges: typeof changes = {};
+    Object.assign(realChanges, changes);
+    // we don't need id, it's not a change, it's the id of the data we want to change
+    if (props.includes('id')) {
+      delete realChanges.id;
+    }
+    for (const prop of props) {
+      if (realChanges[prop] === undefined) {
+        // TODO: maybe a better way than circumventing dynamic delete? I don't see a way around
+        //     deleting dynamically right now.
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+        delete realChanges[prop];
+        continue;
+      }
+      const unwrapped = unwrapSecured(oldObject[prop]);
+      if (unwrapped instanceof Array) {
+        if (isEqual(sortBy(unwrapped), sortBy(realChanges[prop]))) {
+          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+          delete realChanges[prop];
+        }
+      }
+
+      if (unwrapped instanceof Date) {
+        if (unwrapped.getTime() === (realChanges[prop] as Date).getTime()) {
+          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+          delete realChanges[prop];
+          continue;
+        }
+      }
+      if (unwrapped instanceof DateTime) {
+        if (unwrapped.equals(realChanges[prop] as DateTime)) {
```

Comment:
Either of these could be null. We also don't want to use .equals() because that checks if both sides have the same timezone which we don't care about. You can use `a.valueOf() === b.valueOf()`, after checking both are DateTimes.